### ISSUE TITLE
[Feature request] Cache search operations

### ISSUE THREAD

**Issue by Nick-Mazuk** [typesense/typesense-js#52]
With several other search clients (e.g., Algolia's), search results are cached on the client. This way, users don't need to perform unnecessary search requests (e.g., when deleting the last character in their query). This also reduces latency for repeated searches to 0.

I've personally implemented this on my end with the following pattern:

```js
const cachedSearches = {}

export const search = async (options) => {
    const optionsString = JSON.stringify(options)
    if (typeof cachedSearches[optionsString] !== 'undefined') return cachedSearches[optionsString]
    const response = await client.collection('collection').documents.search(options)
    // if statement reduces possible race conditions
    if (typeof cachedSearches[optionsString] === 'undefined') cachedSearches[optionsString] = response
    return response
}
```

My full code is a bit more complex, mainly making sure edge cases (e.g., options appearing in a different order) are still cached correctly. But this is the main idea.

Under the hood, Typesense could use the final search URL as the keys for the cached searches. That way, all collections can be cached with the same cache.

**Comment by jasonbosco** [comment]
Ah yes, query caching has been in the back of my mind to implement. I was thinking of adding it in the instantsearch adapter though, but now that you bring it up, I can see it being useful in the JS library as well.

**Comment by jasonbosco** [comment]
Hi @Nick-Mazuk, I just published [`v0.13.0-0`](https://www.npmjs.com/package/typesense/v/0.13.0-0) with support for search results caching. It's opt-in though to avoid any surprises, especially in dev env. 

There's a new configuration parameter called `cacheSearchResultsForSeconds` which defaults to 0 (disabled). Setting it to a non-zero value will enable caching. Here's an example of how to use it:

https://github.com/typesense/typesense-js/blob/59c7bba882f878f54b786e95a3242bc8c5e6a7b5/test/Typesense/Documents.spec.js#L21-L29

In the [InstantSearch.js adapter](https://github.com/typesense/typesense-instantsearch-adapter) though, I've [set it to cache](https://github.com/typesense/typesense-instantsearch-adapter/pull/42/commits/aaa447a6e6d48e56011ecd0ec598f1168aeb2ba1#diff-388a1dd8bcadb7744c4274d339c43bed1d089c852c615ca970fd4482cfc6c290R16-R18) results for 2 minutes by default as of [`v1.2.1-6`](https://www.npmjs.com/package/typesense-instantsearch-adapter/v/1.2.1-6)

**Comment by Nick-Mazuk** [comment]
Thanks @jasonbosco! I'll test that out when I get a chance (hopefully soon).
